### PR TITLE
[SPARK-42190][K8S] Support `local` mode in `spark.kubernetes.driver.master`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to support a driver-only K8s Spark job.

### Why are the changes needed?

Some workloads like SQL DDL operations can be executed by driver only.

### Does this PR introduce _any_ user-facing change?

No. This is a new feature.

### How was this patch tested?

Pass the CIs with the newly add test case.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this PR, please, @viirya ?

File: docs/running-on-kubernetes.md [hunk 1]
```diff
@@ -590,7 +590,8 @@ See the [configuration page](configuration.html) for information on Spark config
   <td><code>spark.kubernetes.driver.master</code></td>
   <td><code>https://kubernetes.default.svc</code></td>
   <td>
-    The internal Kubernetes master (API server) address to be used for driver to request executors.
+    The internal Kubernetes master (API server) address to be used for driver to request executors or
+    'local[*]' for driver-pod-only mode.
```
Comment (viirya) [inline docs/running-on-kubernetes.md:594]: Is local master URL format all supported? E.g, `local[K,F]`?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala [hunk 2]
```diff
@@ -465,10 +465,12 @@ class KubernetesSuite extends SparkFunSuite
...
-    // at some point.
-    Eventually.eventually(TIMEOUT, patienceInterval) {
-      execPods.values.nonEmpty should be (true)
+    if (patienceInterval.value.toSeconds.toInt > 0) {
+      // If we're testing decommissioning we an executors, but we should have an executor
+      // at some point.
+      Eventually.eventually(TIMEOUT, patienceInterval) {
+        execPods.values.nonEmpty should be (true)
+      }
     }
```
Comment (viirya) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:474]: Is this related?

**Review by viirya — APPROVED** [review]
Looks good to me. A few questions.

File: docs/running-on-kubernetes.md (hunk 1 above)
Comment (HyukjinKwon) [reply docs/running-on-kubernetes.md:594]: Good point. I think we should probably leverage the same regex done in `SparkContextcreateTaskScheduler`, e.g., `LOCAL_N_REGEX`.

Comment (dongjoon-hyun) [reply docs/running-on-kubernetes.md:594]: Sure, we can do that.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:474]: Yes, previously, integration test suite assume the existence of executors. So, without this change, new test case fails after running succesfully.

**Review by HyukjinKwon — APPROVED** [review]
I am fine as is, and doing https://github.com/apache/spark/pull/39748/files#r1087368805 separately. will leave it to you, LGTM

**Comment by viirya** [comment]
Yea, it could be a follow up work.

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya and @HyukjinKwon . I addressed the comments too.

**Comment by dongjoon-hyun** [comment]
If you don't mind, I want to deliver this to Apache Spark 3.4 as a counter-part of [SPARK-41550 Dynamic Allocation on K8S GA](https://issues.apache.org/jira/browse/SPARK-41550), @viirya and @HyukjinKwon . As a zero-executor case, this will complete the Spark implementation in the dimension of the number of executors.

**Comment by viirya** [comment]
I'm fine with this in 3.4.

**Comment by dongjoon-hyun** [comment]
Thank you so much!

**Comment by HyukjinKwon** [comment]
Yeah, I am fine too :-)

**Comment by dongjoon-hyun** [comment]
Thank you all. Merged to master/3.4.
